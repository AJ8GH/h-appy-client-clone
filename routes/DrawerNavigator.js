import React from 'react';
import {useState, useEffect } from 'react';
import { StyleSheet } from 'react-native';
import { NavigationContainer, Button } from '@react-navigation/native';
import { createDrawerNavigator, DrawerContentScrollView, DrawerItemList, DrawerItem } from '@react-navigation/drawer';
import MainMenu from '../views/MainMenu';
import About from '../views/About';
import IndividualCourse from '../views/IndividualCourse';
import AddActivity from '../views/AddActivity';
import {
  NIBBLES, APPETISERS, MAINS, DESSERTS,
} from '../src/CourseDescriptions';
import SignUp from '../views/SignUp';
import SignIn from '../views/SignIn';
import AuthContext from '../src/AuthContext';





const Drawer = createDrawerNavigator();

function Nibbles() {
  return (
    <IndividualCourse
      header="Nibbles"
      dataKey="nibbles"
      description={NIBBLES}
    />
  );
}

function Starters() {
  return (
    <IndividualCourse
      header="Starters"
      dataKey="appetisers"
      description={APPETISERS}
    />
  );
}

function Mains() {
  return (
    <IndividualCourse
      header="Mains"
      dataKey="mains"
      description={MAINS}
    />
  );
}

function Desserts() {
  return (
    <IndividualCourse
      header="Desserts"
      dataKey="desserts"
      description={DESSERTS}
    />
  );
}

function SignOutButton(props) {
  const { signOut } = React.useContext(AuthContext);
  return (
    <DrawerContentScrollView {...props}>
      <DrawerItemList {...props} />
      <DrawerItem {...props}
        label="Sign Out"
        onPress={() => signOut()}
        style={styles.signout}
      />
    </DrawerContentScrollView>
  )
}

const DrawerNavigator = ({state}) => {
  if (state.userToken) {
    return (
      <Drawer.Navigator 
        drawerStyle={styles.drawer}
        drawerContent={(props) => <SignOutButton {...props} />}
        drawerContentOptions={drawerContentStyles}
        >
        <Drawer.Screen name="Menu" component={MainMenu} />
        <Drawer.Screen name="Nibbles" component={Nibbles} />
        <Drawer.Screen name="Starters" component={Starters} />
        <Drawer.Screen name="Mains" component={Mains} />
        <Drawer.Screen name="Desserts" component={Desserts} />
        <Drawer.Screen name="Add activity" component={AddActivity} />
        </Drawer.Navigator>
        )
      } else {
     return (
      <Drawer.Navigator 
        drawerStyle={styles.drawer}
        drawerContentOptions={drawerContentStyles}
        >
        <Drawer.Screen name="Menu" component={MainMenu} />
        <Drawer.Screen name="Nibbles" component={Nibbles} />
        <Drawer.Screen name="Starters" component={Starters} />
        <Drawer.Screen name="Mains" component={Mains} />
        <Drawer.Screen name="Desserts" component={Desserts} />
        <Drawer.Screen name="Sign up" component={SignUp} />
        <Drawer.Screen name="Sign in" component={SignIn} />
        </Drawer.Navigator>
        )
     }
    };

export default DrawerNavigator;

const styles = StyleSheet.create({
  drawer: {
    backgroundColor: '#363946',
    width: 240,
  },
  signout: {
    fontFamily: 'Didot',
    fontSize: 20,
    paddingTop: 10,
    color: '#B1B6A6',
  }
});

const drawerContentStyles = {
  activeTintColor: '#819595',
  itemStyle: {
    marginVertical: 5,
  },
  labelStyle: {
    fontFamily: 'Didot',
    fontSize: 20,
    paddingTop: 10,
    color: '#B1B6A6',
  },
};
