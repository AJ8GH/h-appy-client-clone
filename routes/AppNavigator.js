import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import MainMenu from '../views/MainMenu';
import About from '../views/About';
import IndividualCourse from '../views/IndividualCourse';
import AddActivity from '../views/AddActivity';
import IndividualActivity from '../views/IndividualActivity';
import DrawerNavigator from './DrawerNavigator';

const { Navigator, Screen } = createStackNavigator();

const HomeNavigator = ({state}) => {
    return (
  <Navigator headerMode="none">
    <Screen name="Drawer" children={()=><DrawerNavigator state={state}/>}/>
    <Screen name="IndividualActivity" component={IndividualActivity} />
  </Navigator>
)};

const AppNavigator = ({state}) => {
  return (
  <NavigationContainer>
    <HomeNavigator state={state}/>
  </NavigationContainer>
)};

export default AppNavigator;
