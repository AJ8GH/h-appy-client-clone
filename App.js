/* eslint-disable no-use-before-define */
import React from 'react';
import fetch from 'node-fetch';
import { useState, useEffect, useContext } from 'react';
import {
  StyleSheet, Platform, SafeAreaView, ScrollView, StatusBar,
} from 'react-native';
import { showMessage, hideMessage } from "react-native-flash-message";
import AppNavigator from './routes/AppNavigator';
import DrawerNavigator from './routes/DrawerNavigator';
import FlashMessage from 'react-native-flash-message';
import SecureStore from 'expo-secure-store';
import sendAuthenticationData from './src/AuthenticationData';
import AuthContext from './src/AuthContext';

export default function App({navigation}) {
  const [state, dispatch] = React.useReducer(
    (prevState, action) => {
      switch (action.type) {
        case 'RESTORE_TOKEN':
          return {
            ...prevState,
            userToken: action.token,
            isLoading: false,
          };
        case 'SIGN_IN':
          return {
            ...prevState,
            isSignout: false,
            userToken: action.token,
          };
        case 'SIGN_OUT':
          return {
            ...prevState,
            isSignout: true,
            userToken: null,
          };
      }
    },
    {
      isLoading: true,
      isSignout: false,
      userToken: null,
    }
  );

  React.useEffect(() => {
    const bootstrapAsync = async () => {
      let userToken;

      try {
        userToken = await SecureStore.getItemAsync('userToken');
      } catch (error) {

      }
      dispatch({ type: 'RESTORE_TOKEN', token: userToken || null });
    };

    bootstrapAsync();
  }, []);

  const authContext = React.useMemo(
    () => ({
      signIn: async (emailData, passwordData, navigation) => {
        let response = await fetch('http://localhost:3000/user/login', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email: emailData,
          password: passwordData,
        })
      });
        let json = await response.json();
        if (json.user) {
          showMessage({
            message: "Sign in successful",
            description: `Welcome back to H-Appy, ${json.user.name}!`,
            type: "success",
          });
          dispatch({ type: 'SIGN_IN', token: JSON.stringify(json.token)})
          navigation.navigate('Menu')
        } else {
           showMessage({
             message: "Authentication failed",
             description: "Those details don't match our records",
             type: "error",
           });
         };
      },
      signOut: async (navigation, userToken) => {
        await fetch('http://localhost:3000/user/profile/logout', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ' + state.userToken
        },
      });
        dispatch({ type: 'SIGN_OUT'})
        navigation.navigate('Menu')
        showMessage({
          message: "Sign out successful",
          description: 'Come back soon!',
          type: "success",
        })
    },
      signUp: async (nameData, usernameData, emailData, passwordData, navigation) => {
        let response = await fetch('http://localhost:3000/user', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            name: nameData,
            username: usernameData,
            email: emailData,
            password: passwordData,
          })
        });
        let json = await response.json();
        if (json.user) {
          showMessage({
            message: "Signup successful",
            description: `Welcome to H-Appy, ${json.user.name}!`,
            type: "success",
          });
          dispatch({ type: 'SIGN_IN', token: JSON.stringify(json.token)})
          navigation.navigate('Menu')
        } else if (json.name && json.name === "MongoError") {
          if ("email" in json.keyPattern) {
            showMessage({
              message: "Email in use",
              description: "That email has already been taken",
              type: "error",
            });
          }
          else if ("username" in json.keyPattern) {
            showMessage({
              message: "Username in use",
              description: "That username has already been taken",
              type: "error",
            });
          }
        }
        else if (json.errors) {
          if (json.errors.email && json.errors.email.name === "ValidatorError") {
            showMessage({
              message: "Email not valid",
              description: "You must submit a valid email",
              type: "error",
            });
          }
          else if (json.errors.password && json.errors.password.name === "ValidatorError") {
            showMessage({
              message: "Password too short",
              description: "Your password must be at least 8 characters long",
              type: "error",
            });
          }
        }
      },
    }),
    [],
  );

  console.log(state.userToken)

  return (
    <>
    <AuthContext.Provider value={authContext}>
      <SafeAreaView style={styles.safeAreaTop} />
      <SafeAreaView style={styles.safeAreaBottom}>
        <FlashMessage position="top" />
        <AppNavigator state = {state}/>
      </SafeAreaView>
      <StatusBar barStyle="light-content" />
      </AuthContext.Provider>
    </>
  );
}

const styles = StyleSheet.create({
  safeAreaTop: {
    flex: 0,
    backgroundColor: '#363946',
    paddingTop: Platform.OS === 'android' ? 25 : 0,
  },
  safeAreaBottom: {
    flex: 1,
    backgroundColor: '#819595',
  },
});

const validation = (response, navigation) => {
  if (response.user) {
    showMessage({
      message: "Signup successful",
      description: `Welcome to H-Appy, ${response.user.name}!`,
      type: "success",
    });
    dispatch({ type: 'SIGN_IN', token: JSON.stringify(response.token)})
    navigation.navigate('Menu')
  } else if (response.name && response.name === "MongoError") {
   if ("email" in response.keyPattern) {
     showMessage({
       message: "Email in use",
       description: "That email has already been taken",
       type: "error",
     });
   }
   else if ("username" in response.keyPattern) {
     showMessage({
       message: "Username in use",
       description: "That username has already been taken",
       type: "error",
     });
   }
 }
 else if (response.errors) {
   if (response.errors.email && response.errors.email.name === "ValidatorError") {
     showMessage({
       message: "Email not valid",
       description: "You must submit a valid email",
       type: "error",
     });
   }
   else if (response.errors.password && response.errors.password.name === "ValidatorError") {
     showMessage({
       message: "Password too short",
       description: "Your password must be at least 8 characters long",
       type: "error",
     });
   }
 }
}