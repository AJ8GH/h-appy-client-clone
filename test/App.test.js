import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App';
import DrawerNavigator from '../routes/DrawerNavigator'
import { Platform } from 'react-native';
import AppNavigator from '../routes/AppNavigator';

jest.mock('@react-navigation');
AppNavigator.mockImplementation();

jest.mock('../routes/AppNavigator');
AppNavigator.mockImplementation();

jest.mock('react-native');
Platform.mockImplementation(() => {OS: 'iOS'});

jest.mock('../routes/DrawerNavigator');
DrawerNavigator.mockImplementation(() => 'DrawerNavigator mocked');

describe('App', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})
