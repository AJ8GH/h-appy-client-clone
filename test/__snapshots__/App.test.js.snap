// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f8f9d4",
      "flex": 1,
      "paddingTop": 0,
    }
  }
>
  DrawerNavigator mocked
</RCTSafeAreaView>
`;
